src/Services/LoopService.cs

```diff
@@ -0,0 +1,208 @@
+    {
+        _logger = logger;
+    }
+
+    public GrpcChannel CreateClient(string endpoint)
+    {
+        if (string.IsNullOrWhiteSpace(endpoint))
+        {
+            throw new ArgumentException("Endpoint cannot be null or empty.", nameof(endpoint));
+        }
+
+        var handler = new HttpClientHandler
+        {
+            // Accept any server certificate for development purposes
+            ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+        };
+
+        // Create a new gRPC channel for the specified endpoint
+        var channel = GrpcChannel.ForAddress($"https://{endpoint}",
+            new GrpcChannelOptions
+            {
+                HttpHandler = handler,
+                LoggerFactory = NullLoggerFactory.Instance
+            });
+
+        _logger.LogInformation("Creating gRPC client for Loop server at {Endpoint}", endpoint);
+
+        return channel;
+    }
+    
+    public SwapClient.SwapClientClient GetClient(Node node)
+    {
+        if (string.IsNullOrEmpty(node.LoopdEndpoint) || string.IsNullOrEmpty(node.LoopdMacaroon))
+        {
+            throw new ArgumentException("Node Loopd endpoint or macaroon is not set.");
+        }
+
+        var clientKey = $"{node.LoopdEndpoint}";
+        return _clients.GetOrAdd(clientKey, _ =>
+        {
+            var channel = _channels.GetOrAdd(node.LoopdEndpoint, CreateClient(node.LoopdEndpoint));
+            var invoker = channel.Intercept(new GRPCMacaroonInterceptor(node.LoopdMacaroon));
+            return new SwapClient.SwapClientClient(invoker);
+        });
+    }
+
+    public async Task<bool> PingAsync(Node node, CancellationToken cancellationToken= default)
+    {
+        if (string.IsNullOrEmpty(node.LoopdEndpoint) || string.IsNullOrEmpty(node.LoopdMacaroon))
+        {
+            throw new ArgumentException("Node Loopd endpoint or macaroon is not set.");
+        }
+
+        var client = GetClient(node);
+
+        var response = await client.GetInfoAsync(new GetInfoRequest(), null, null, cancellationToken);
+        if (string.IsNullOrEmpty(response.Version))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    public async Task<SwapResponse> CreateSwapOutAsync(Node node, SwapOutRequest request, CancellationToken cancellationToken= default)
+    {
+        if (string.IsNullOrEmpty(node.LoopdEndpoint) || string.IsNullOrEmpty(node.LoopdMacaroon))
+        {
+            throw new ArgumentException("Node Loopd endpoint or macaroon is not set.");
+        }
+
+        var client = GetClient(node);
+
+        const long FeeRateTotalParts = 1_000_000;
+        // Define route independent max routing fees. We have currently no way
+        // to get a reliable estimate of the routing fees. Best we can do is
+        // the minimum routing fees, which is not very indicative.
+        var maxRoutingFeeBase = 10;
+
+        var maxRoutingFeeRate = 20000;
```

Comment:
use user %

Reply:
added clarifying comments to make it easier to come back to